In [1]:
from pathlib import Path
import pandas as pd

csv_path = Path('data/raw/ld50-smiles-descriptors-dataset.csv')
if not csv_path.is_file():
    csv_path = Path('../data/raw/ld50-smiles-descriptors-dataset.csv')

df = pd.read_csv(csv_path)

In [2]:
df.head()

,Name,SMILES,LD50,BertzCT,Chi2v,Chi3n,Chi3v,Chi4n,Chi4v,HeavyAtomCount,...,NumHAcceptors,NumHDonors,NumHeteroatoms,RingCount,SMR_VSA10,SlogP_VSA12,SlogP_VSA5,VSA_EState4,VSA_EState9,qed
0,4-nitroaniline,Nc1ccc([N+](=O)[O-])cc1,2.265,242.429658,1.848642,1.103907,1.103907,0.596038,0.596038,10,...,3,1,4,1,11.374773,0.00000,0.000000,5.901255,0.000000,0.359544
1,4-nitrophenol,O=[N+]([O-])c1ccc(O)cc1,2.838,241.674771,1.773508,1.060528,1.060528,0.574349,0.574349,10,...,3,1,4,1,5.687386,0.00000,0.000000,-0.015926,0.000000,0.470728
2,1-(chloromethyl)-4-nitrobenzene,O=[N+]([O-])c1ccc(CCl)cc1,1.977,257.648013,2.324449,1.301462,1.610069,0.706751,0.884925,11,...,2,0,4,1,17.288326,11.60094,5.563451,0.982531,0.000000,0.389482
3,1-methoxy-4-nitrobenzene,COc1ccc([N+](=O)[O-])cc1,1.823,254.893126,1.955135,1.283242,1.283242,0.703937,0.703937,11,...,3,0,4,1,5.687386,0.00000,0.000000,0.074815,1.515046,0.478623
4,"benzene-1,4-dicarbonyl chloride",O=C(Cl)c1ccc(C(=O)Cl)cc1,1.910,283.334675,2.811331,1.338820,1.775256,0.732299,0.984275,12,...,2,0,4,1,33.686733,23.20188,20.715977,0.691358,0.000000,0.691354


In [3]:
df.shape

(7397, 28)

In [4]:
df.columns

Index(['Name', 'SMILES', 'LD50', 'BertzCT', 'Chi2v', 'Chi3n', 'Chi3v', 'Chi4n',
       'Chi4v', 'HeavyAtomCount', 'HeavyAtomMolWt', 'Kappa3',
       'MaxPartialCharge', 'MinAbsPartialCharge', 'MinEStateIndex',
       'MinPartialCharge', 'MolLogP', 'MolWt', 'NumHAcceptors', 'NumHDonors',
       'NumHeteroatoms', 'RingCount', 'SMR_VSA10', 'SlogP_VSA12', 'SlogP_VSA5',
       'VSA_EState4', 'VSA_EState9', 'qed'],
      dtype='str')

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7397 entries, 0 to 7396
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Name                 7397 non-null   str    
 1   SMILES               7397 non-null   str    
 2   LD50                 7396 non-null   float64
 3   BertzCT              7392 non-null   float64
 4   Chi2v                7389 non-null   float64
 5   Chi3n                7389 non-null   float64
 6   Chi3v                7389 non-null   float64
 7   Chi4n                7393 non-null   float64
 8   Chi4v                7394 non-null   float64
 9   HeavyAtomCount       7397 non-null   int64  
 10  HeavyAtomMolWt       7396 non-null   float64
 11  Kappa3               7394 non-null   float64
 12  MaxPartialCharge     7392 non-null   float64
 13  MinAbsPartialCharge  7391 non-null   float64
 14  MinEStateIndex       7393 non-null   float64
 15  MinPartialCharge     7391 non-null   float64
 16 

In [6]:
df.describe()

,LD50,BertzCT,Chi2v,Chi3n,Chi3v,Chi4n,Chi4v,HeavyAtomCount,HeavyAtomMolWt,Kappa3,...,NumHAcceptors,NumHDonors,NumHeteroatoms,RingCount,SMR_VSA10,SlogP_VSA12,SlogP_VSA5,VSA_EState4,VSA_EState9,qed
count,7396.00000,7392.000000,7389.000000,7389.000000,7389.000000,7393.000000,7394.000000,7397.000000,7396.000000,7394.000000,...,7397.000000,7397.000000,7397.000000,7397.000000,7397.000000,7397.000000,7397.000000,7397.000000,7397.000000,7397.000000
mean,2.55419,400.140968,4.908541,2.473274,3.366371,1.622177,2.282153,16.169258,232.049389,7.799745,...,3.299851,0.787076,4.743409,1.417331,18.276877,9.552595,20.621943,1.267097,0.333946,0.567852
std,0.95169,305.446876,2.950234,1.832664,2.459651,1.436304,2.052736,7.705312,109.232519,160.977491,...,2.286146,1.163119,2.917261,1.313732,17.292006,15.629996,21.516070,2.627986,1.930950,0.173972
min,0.29100,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.000000,26.018000,-27.040000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,-15.221693,-19.025978,0.032375
25%,1.86700,169.256513,2.784621,1.308924,1.651111,0.724270,0.944832,11.000000,152.108000,2.058290,...,2.000000,0.000000,3.000000,0.000000,5.969305,0.000000,5.563451,0.000000,0.000000,0.446898
50%,2.37800,351.572890,4.318588,2.050387,2.809473,1.259558,1.785409,15.000000,218.183500,3.275854,...,3.000000,0.000000,4.000000,1.000000,12.310731,0.000000,13.847474,0.311944,0.000000,0.564131
75%,3.04400,552.137280,6.462885,3.171446,4.516022,2.122591,3.046399,20.000000,292.240250,4.856951,...,4.000000,1.000000,6.000000,2.000000,27.293868,12.217873,28.763083,1.883273,0.000000,0.701040
max,7.20700,4025.484341,26.092602,21.050990,23.784903,16.235218,31.500000,90.000000,1247.715000,9507.960000,...,35.000000,21.000000,35.000000,21.000000,182.425200,139.211279,208.079042,29.904342,23.150432,0.944994


In [7]:
df.isnull().sum()

Name                   0
SMILES                 0
LD50                   1
BertzCT                5
Chi2v                  8
Chi3n                  8
Chi3v                  8
Chi4n                  4
Chi4v                  3
HeavyAtomCount         0
HeavyAtomMolWt         1
Kappa3                 3
MaxPartialCharge       5
MinAbsPartialCharge    6
MinEStateIndex         4
MinPartialCharge       6
MolLogP                5
MolWt                  5
NumHAcceptors          0
NumHDonors             0
NumHeteroatoms         0
RingCount              0
SMR_VSA10              0
SlogP_VSA12            0
SlogP_VSA5             0
VSA_EState4            0
VSA_EState9            0
qed                    0
dtype: int64

In [8]:
df.duplicated().sum()

np.int64(8)

## Clean and prepare the data

Work on a copy of `df`. `LD50` is already stored as −log10(LD50 mol/kg), so it is not transformed. Only missing target values and exact duplicate rows are removed.

### 1. Create a working copy

In [9]:
# Copy the data so the original stays unchanged.
clean_df = df.copy()
# Record the starting number of rows.
initial_rows = len(clean_df)
# Choose the two columns to check.
main_columns = ['LD50', 'NumHAcceptors']

### 2. Check missing target values

In [10]:
# Count missing values in each selected column.
missing_values = clean_df[main_columns].isnull().sum()
# Display the missing-value counts.
missing_values

LD50             1
NumHAcceptors    0
dtype: int64

### 3. Remove rows missing LD50 or NumHAcceptors

In [11]:
# Remove rows missing LD50 or NumHAcceptors.
clean_df = clean_df.dropna(subset=['LD50', 'NumHAcceptors'])
# Calculate how many rows were removed.
missing_rows_removed = initial_rows - len(clean_df)
# Print the number of removed rows.
print(f'Rows removed for missing target values: {missing_rows_removed}')
# Check that neither selected column has missing values left.
assert not clean_df[main_columns].isnull().any().any()
# Display the remaining missing-value counts.
clean_df[main_columns].isnull().sum()

Rows removed for missing target values: 1


LD50             0
NumHAcceptors    0
dtype: int64

### 4. Check exact duplicate rows

In [12]:
# Count rows that exactly repeat an earlier row.
exact_duplicate_rows = int(clean_df.duplicated().sum())
# Display the number of exact duplicate rows.
exact_duplicate_rows

8

### 5. Check duplicate SMILES separately

Repeated SMILES may represent separate toxicity records, so SMILES alone is not used to remove rows.

In [13]:
# Count SMILES entries that repeat an earlier value.
duplicate_smiles_before = int(clean_df['SMILES'].duplicated().sum())
# Display the number of repeated SMILES entries.
duplicate_smiles_before

55

### 6. Remove exact duplicates

Keep the first occurrence only when every column is identical. Retain other records that share a SMILES value.

In [14]:
# Record the row count before removing duplicates.
rows_before_deduplication = len(clean_df)
# Run the removal only if exact duplicates exist.
if exact_duplicate_rows > 0:
    # Keep the first copy of each identical row.
    clean_df = clean_df.drop_duplicates()
# Calculate how many duplicate rows were removed.
exact_rows_removed = rows_before_deduplication - len(clean_df)
# Check that no exact duplicate rows remain.
assert not clean_df.duplicated().any()
# Print the number of duplicate rows removed.
print(f'Exact duplicate rows removed: {exact_rows_removed}')
# Print how many repeated SMILES entries remain.
print(f'Remaining repeated SMILES values: {clean_df["SMILES"].duplicated().sum()}')

Exact duplicate rows removed: 8
Remaining repeated SMILES values: 47


### 7. Confirm numeric values

Check that both target columns are numeric and finite. Stop for review if validation fails instead of silently changing values.

In [15]:
# Import NumPy to check for missing or infinite numbers.
import numpy as np
# Import a helper that checks whether a data type is numeric.
from pandas.api.types import is_numeric_dtype

# Check whether each selected column has a numeric data type.
numeric_columns = clean_df[main_columns].dtypes.apply(is_numeric_dtype)
# Stop with an error if either column is not numeric.
assert numeric_columns.all(), 'LD50 and NumHAcceptors must both be numeric.'
# Check that both columns contain no missing or infinite numbers.
assert np.isfinite(clean_df[main_columns]).all().all(), 'Review non-finite target values.'
# Display the data type of each selected column.
clean_df[main_columns].dtypes

LD50             float64
NumHAcceptors      int64
dtype: object

### 8. Validate hydrogen-bond acceptor counts

In [16]:
# Select the hydrogen-bond acceptor counts.
acceptor_counts = clean_df['NumHAcceptors']
# Mark counts that are whole numbers and zero or greater.
valid_acceptor_counts = acceptor_counts.ge(0) & acceptor_counts.mod(1).eq(0)
# Print how many counts failed these checks.
print(f'Invalid acceptor counts: {(~valid_acceptor_counts).sum()}')
# Check that every acceptor count is valid.
assert valid_acceptor_counts.all(), 'Review negative or fractional NumHAcceptors values.'
# Show how often each count occurs, from smallest to largest.
clean_df['NumHAcceptors'].value_counts().sort_index()

Invalid acceptor counts: 0


NumHAcceptors
0      287
1     1172
2     1653
3     1461
4     1110
5      655
6      470
7      251
8      173
9       46
10      29
11      14
12      18
13      16
14      11
15       7
16       4
17       1
18       2
19       4
23       1
24       1
25       1
35       1
Name: count, dtype: int64

### 9. Inspect extreme values

Inspect ranges, tail percentiles, and the five lowest and highest records for each target column. Unusual values are retained; these checks do not establish that an extreme value is an error.

In [17]:
# Show summary statistics, including the 1st, 50th, and 99th percentiles.
clean_df[main_columns].describe(percentiles=[0.01, 0.5, 0.99])

,LD50,NumHAcceptors
count,7388.000000,7388.000000
mean,2.553692,3.299675
std,0.951543,2.286698
min,0.291000,0.000000
1%,0.955480,0.000000
50%,2.378000,3.000000
99%,5.242430,11.000000
max,7.207000,35.000000


In [18]:
# Import the function used to show tables in the notebook.
from IPython.display import display

# Choose which columns to show for each record.
inspection_columns = ['Name', 'SMILES', 'LD50', 'NumHAcceptors']
# Inspect LD50 and NumHAcceptors one at a time.
for column in main_columns:
    # Print a heading for the lowest values.
    print(f'Five lowest {column} values:')
    # Display the five rows with the lowest values.
    display(clean_df.nsmallest(5, column)[inspection_columns])
    # Print a heading for the highest values.
    print(f'Five highest {column} values:')
    # Display the five rows with the highest values.
    display(clean_df.nlargest(5, column)[inspection_columns])

Five lowest LD50 values:


,Name,SMILES,LD50,NumHAcceptors
6971,"6-methyl-1h-pyrimidine-2,4-dione",Cc1cc(=O)[nH]c(=O)[nH]1,0.291,2
402,"4-methyl-1,3-dioxolan-2-one",CC1COC(=O)O1,0.470,3
496,hexane,CCCCCC,0.537,0
6151,"bis(2-ethylhexyl) cyclohexene-1,2-dicarboxylate",CCCCC(CC)COC(=O)C1=C(C(=O)OCC(CC)CCCC)CCCC1,0.539,4
2393,2-(cyclohexylamino)ethanol,OCCNC1CCCCC1,0.573,2


Five highest LD50 values:


,Name,SMILES,LD50,NumHAcceptors
6271,"2,3,7,8-tetrachlorodibenzo-p-dioxin",Clc1cc2c(cc1Cl)Oc1cc(Cl)c(Cl)cc1O2,7.207,2
5348,"(5z)-5-[(2e)-2-[7a-methyl-1-[7,7,7-trifluoro-6...",C=C1/C(=C\C=C2/CCCC3(C)C2CCC3C(C)CCCC(O)(C(F)(...,7.100,3
3480,"2,3,7,8-tetrabromodibenzo-p-dioxin",Brc1cc2c(cc1Br)Oc1cc(Br)c(Br)cc1O2,6.699,2
6854,"3-[3-[4-(4-bromophenyl)phenyl]-1,2,3,4-tetrahy...",O=c1oc2ccccc2c(O)c1C1CC(c2ccc(-c3ccc(Br)cc3)cc...,6.515,3
1956,3-ethyl-n-[methoxy(methylsulfanyl)phosphoryl]-...,CCN1CC(C)O/C1=N\P(=O)(OC)SC,6.402,4


Five lowest NumHAcceptors values:


,Name,SMILES,LD50,NumHAcceptors
9,4-ethenylcyclohexene,C=CC1CC=CCC1,1.626,0
10,ethylbenzene,CCc1ccccc1,1.482,0
11,styrene,C=Cc1ccccc1,1.594,0
13,chloromethylbenzene,ClCc1ccccc1,2.012,0
79,1-chloro-4-[(4-chlorophenyl)methyl]benzene,Clc1ccc(Cc2ccc(Cl)cc2)cc1,2.375,0


Five highest NumHAcceptors values:


,Name,SMILES,LD50,NumHAcceptors
5028,"(1s,3r,5r,6s,8r,10r,11s,13r,15r,16s,18r,20r,21...",OC[C@H]1O[C@@H]2O[C@@H]3[C@@H](CO)O[C@H](O[C@@...,1.781,35
5727,"4-[3-[3,4-dihydroxy-5-(3,4,5-trihydroxybenzoyl...",O=C(Oc1cc(C(=O)Oc2cc(C(=O)OC3C(OC(=O)c4cc(O)c(...,2.411,25
5720,"bis[(2s,3r,4s,5s,6r)-3,4,5-trihydroxy-6-[[(2r,...",CC(/C=C/C=C(\C)C(=O)O[C@@H]1O[C@H](CO[C@@H]2O[...,2.490,24
4711,"6-[[9-acetyloxy-8-hydroxy-4,8a-bis(hydroxymeth...",C/C=C(/C)C(=O)OC1C(OC(C)=O)C2(CO)C(O)CC3(C)C(=...,3.133,23
142,"[3-[5-(4,5-diformyloxy-6-methyloxan-2-yl)oxy-4...",CC1OC(OC2C(C)OC(OC3C(C)OC(OC4CCC5(C)C(CCC6C5CC...,4.373,19


### 10. Save the cleaned dataset

Save all columns without a CSV index. Resolve the output directory from the loaded CSV path so this works from either the project root or `notebooks/`.

In [19]:
# Set the cleaned CSV path inside data/processed.
processed_path = csv_path.parent.parent / 'processed' / 'ld50_cleaned.csv'
# Create the output folder if it does not already exist.
processed_path.parent.mkdir(parents=True, exist_ok=True)
# Save the cleaned data without the row index.
clean_df.to_csv(processed_path, index=False)
# Print the number of saved rows and the file location.
print(f'Saved {len(clean_df):,} rows to {processed_path.as_posix()}')
# Print the total rows removed and the reasons.
print(f'Total rows removed: {initial_rows - len(clean_df)} '
      f'({missing_rows_removed} missing target values; {exact_rows_removed} exact duplicates)')

Saved 7,388 rows to ../data/processed/ld50_cleaned.csv
Total rows removed: 9 (1 missing target values; 8 exact duplicates)


### Cleaning summary

For the current raw dataset, 7,397 rows became **7,388 rows**: **1 row** was removed for missing `LD50` (none were missing `NumHAcceptors`), followed by **8 exact duplicate rows**, for **9 rows removed in total**. There were 55 repeated SMILES values before exact deduplication; the remaining 47 are retained as potentially separate toxicity records.

Both target columns are numeric and finite, with no missing values; all `NumHAcceptors` values are non-negative whole numbers. Observed `LD50` values range from **0.291 to 7.207**, and `NumHAcceptors` from **0 to 35**. Extreme values were inspected and retained, and `LD50` was not transformed.

The final dataset is saved to `data/processed/ld50_cleaned.csv`. The original `df` and raw CSV are unchanged.

## Next step: data visualisation

The six graphs, their Python code and explanatory notes are in [02_data_visualisation.ipynb](02_data_visualisation.ipynb), which reads `data/processed/ld50_cleaned.csv`.